src/app/clusters/joint-fabric-administrator-server/joint-fabric-administrator-server.cpp

```diff
@@ -102,31 +100,82 @@ void MatterJointFabricAdministratorPluginServerInitCallback()
     AttributeAccessInterfaceRegistry::Instance().Register(&gJointFabricAdministratorAttrAccess);
 }
 
-// TODO
+#if CHIP_DEVICE_CONFIG_ENABLE_JOINT_FABRIC
+const FabricInfo * RetrieveCurrentFabric(CommandHandler * aCommandHandler)
+{
+    FabricIndex index = aCommandHandler->GetAccessingFabricIndex();
+    ChipLogDetail(Zcl, "JointFabric: Finding fabric with fabricIndex 0x%x", static_cast<unsigned>(index));
+    return Server::GetInstance().GetFabricTable().FindFabricWithIndex(index);
+}
+
 bool emberAfJointFabricAdministratorClusterICACCSRRequestCallback(
     chip::app::CommandHandler * commandObj, const chip::app::ConcreteCommandPath & commandPath,
     const chip::app::Clusters::JointFabricAdministrator::Commands::ICACCSRRequest::DecodableType & commandData)
 {
     MATTER_TRACE_SCOPE("ICACCSRRequest", "JointFabricAdministrator");
+    ChipLogProgress(Zcl, "JointFabricAdministrator: Received an ICACCSRRequest command");
+
+    auto nonDefaultStatus           = Status::Success;
+    const FabricInfo * fabricInfo   = RetrieveCurrentFabric(commandObj);
+    auto & failSafeContext          = Server::GetInstance().GetFailSafeContext();
+    auto & jointFabricAdministrator = Server::GetInstance().GetJointFabricAdministrator();
+
+    uint8_t buf[Credentials::kMaxDERCertLength];
+    MutableByteSpan icacCsr(buf, Credentials::kMaxDERCertLength);
+    Commands::ICACCSRResponse::Type response;
+
+    // If current fabric is not available, command was invoked over PASE which is not legal
+    VerifyOrExit(fabricInfo != nullptr, nonDefaultStatus = Status::InvalidCommand);
 
-    ChipLogProgress(Zcl, "JointFabricAdministrator: Received a ICACCSRRequest command");
+    VerifyOrExit(failSafeContext.IsFailSafeArmed(commandObj->GetAccessingFabricIndex()),
+                 nonDefaultStatus = Status::FailsafeRequired);
+
+    /* TODO spec.: If the <<ref_FabricTableVendorIdVerificationProcedure, FabricFabric Table Vendor ID Verification Procedure>>
+     * has not been executed against the initiator of this command, the command SHALL fail
+     * with a <<ref_JFVidNotVerified, JfVidNotVerified>> status code SHALL be sent back to the initiator.*/
+
+    VerifyOrExit(jointFabricAdministrator.GetDelegate() != nullptr, nonDefaultStatus = Status::Failure);
+    VerifyOrExit(jointFabricAdministrator.GetDelegate()->GetIcacCsr(icacCsr) == CHIP_NO_ERROR, nonDefaultStatus = Status::Failure);
```

Comment:
The spec has a "If a prior AddICAC command was successfully executed within the fail-safe timer period" check here which I don't see in this code?

Reply:
Right, added the check, thx.